File: server/services/assignmentService.js
const db = require("../db/jsonStore");
const { AssignmentError } = require("../utils/errors");
const activityService = require("./activityService");

const STATUSES = ["Pending", "In Progress", "Completed", "On Hold", "Cancelled"];

function enrichWithTask(assignment) {
  const task = db.table("tasks").find(assignment.task_id);
  return Object.assign({}, assignment, {
    task_title: task ? task.title : null,
    task_description: task ? task.description : null,
    task_priority: task ? task.priority : null,
    task_estimated_hours: task ? task.estimated_hours : null,
  });
}

function enrichWithEmployee(assignment) {
  const employee = db.table("employees").find(assignment.employee_id);
  return Object.assign({}, assignment, {
    employee_name: employee ? `${employee.first_name} ${employee.last_name}` : null,
    employee_code: employee ? employee.employee_code : null,
  });
}

function assignTaskToEmployee(taskId, employeeId, actorId) {
  const task = db.table("tasks").find(taskId);
  if (!task || task.is_deleted) throw new AssignmentError("Task not found", 404);

  const employee = db.table("employees").find(employeeId);
  if (!employee || !employee.is_active) throw new AssignmentError("Employee not found", 404);

  if (
    db.table("task_assignments").findOne(
      (a) => a.task_id === Number(taskId) && a.employee_id === Number(employeeId)
    )
  ) {
    throw new AssignmentError("This task is already assigned to that employee");
  }

  const now = new Date().toISOString();
  const assignment = db.table("task_assignments").insert({
    task_id: Number(taskId),
    employee_id: Number(employeeId),
    status: "Pending",
    completion_percentage: 0,
    remarks: "",
    assigned_at: now,
    updated_at: now,
  });

  activityService.log(actorId, "CREATE", "assignment", assignment.id, null, assignment);
  return assignment;
}

function updateAssignmentStatus(id, fields, actorId) {
  const existing = db.table("task_assignments").find(id);
  if (!existing) throw new AssignmentError("Assignment not found", 404);

  const patch = { updated_at: new Date().toISOString() };

  if (fields.status !== undefined) {
    if (!STATUSES.includes(fields.status)) {
      throw new AssignmentError(`Status must be one of: ${STATUSES.join(", ")}`);
    }
    patch.status = fields.status;
  }
  if (fields.completion_percentage !== undefined) {
    const pct = Number(fields.completion_percentage);
    if (Number.isNaN(pct) || pct < 0 || pct > 100) {
      throw new AssignmentError("completion_percentage must be between 0 and 100");
    }
    patch.completion_percentage = pct;
  }
  if (fields.remarks !== undefined) patch.remarks = fields.remarks;

  if (patch.status === "Completed") patch.completion_percentage = 100;

  const updated = db.table("task_assignments").update(id, patch);
  activityService.log(actorId, "UPDATE", "assignment", id, existing, updated);
  return updated;
}

function removeAssignment(id, actorId) {
  const existing = db.table("task_assignments").find(id);
  if (!existing) throw new AssignmentError("Assignment not found", 404);
  db.table("task_assignments").remove(id);
  activityService.log(actorId, "DELETE", "assignment", id, existing, null);
  return true;
}

function getAssignmentsForEmployee(employeeId) {
  return db
    .table("task_assignments")
    .filter((a) => a.employee_id === Number(employeeId))
    .map(enrichWithTask)
    .sort((a, b) => new Date(b.assigned_at) - new Date(a.assigned_at));
}

function getAssignmentsForTask(taskId) {
  return db
    .table("task_assignments")
    .filter((a) => a.task_id === Number(taskId))
    .map(enrichWithEmployee);
}

function listAllAssignments() {
  return db
    .table("task_assignments")
    .all()
    .map((a) => enrichWithTask(enrichWithEmployee(a)))
    .sort((a, b) => new Date(b.assigned_at) - new Date(a.assigned_at));
}

module.exports = {
  assignTaskToEmployee,
  updateAssignmentStatus,
  removeAssignment,
  getAssignmentsForEmployee,
  getAssignmentsForTask,
  listAllAssignments,
  STATUSES,
};

File: backend/task_management/services/assignment_service.py
from datetime import datetime

from sqlalchemy.orm import joinedload

from ..extensions import db
from ..models import Employee, Task, TaskAssignment
from .activity_service import log as log_activity
from .errors import AssignmentError

STATUSES = ("Pending", "In Progress", "Completed", "On Hold", "Cancelled")


def assign_task_to_employee(task_id, employee_id, actor_id):
    task = db.session.get(Task, task_id)
    if not task or task.is_deleted:
        raise AssignmentError("Task not found", 404)

    employee = db.session.get(Employee, employee_id)
    if not employee or not employee.is_active:
        raise AssignmentError("Employee not found", 404)

    if TaskAssignment.query.filter_by(task_id=task_id, employee_id=employee_id).first():
        raise AssignmentError("This task is already assigned to that employee")

    assignment = TaskAssignment(task_id=task_id, employee_id=employee_id)
    db.session.add(assignment)
    db.session.commit()

    log_activity(actor_id, "CREATE", "assignment", assignment.id, None, assignment.to_dict())
    return assignment


def get_assignment(assignment_id):
    assignment = db.session.get(TaskAssignment, assignment_id)
    if not assignment:
        raise AssignmentError("Assignment not found", 404)
    return assignment


def update_assignment_status(assignment_id, fields, actor_id):
    assignment = get_assignment(assignment_id)

    before = assignment.to_dict()

    if fields.get("status") is not None:
        if fields["status"] not in STATUSES:
            raise AssignmentError(f"Status must be one of: {', '.join(STATUSES)}")
        assignment.status = fields["status"]

    if fields.get("completion_percentage") is not None:
        try:
            pct = int(fields["completion_percentage"])
        except (TypeError, ValueError):
            raise AssignmentError("completion_percentage must be a whole number") from None
        if pct < 0 or pct > 100:
            raise AssignmentError("completion_percentage must be between 0 and 100")
        assignment.completion_percentage = pct

    if fields.get("remarks") is not None:
        assignment.remarks = fields["remarks"]

    if assignment.status == "Completed":
        assignment.completion_percentage = 100

    assignment.updated_at = datetime.utcnow()
    db.session.commit()

    log_activity(actor_id, "UPDATE", "assignment", assignment.id, before, assignment.to_dict())
    return assignment


def remove_assignment(assignment_id, actor_id):
    assignment = get_assignment(assignment_id)

    before = assignment.to_dict()
    db.session.delete(assignment)
    db.session.commit()

    log_activity(actor_id, "DELETE", "assignment", assignment_id, before, None)
    return True


def get_assignments_for_employee(employee_id):
    assignments = (
        TaskAssignment.query.options(joinedload(TaskAssignment.task))
        .filter_by(employee_id=employee_id)
        .order_by(TaskAssignment.assigned_at.desc())
        .all()
    )
    return [a.to_dict_with_task() for a in assignments]


def get_assignments_for_task(task_id):
    assignments = (
        TaskAssignment.query.options(joinedload(TaskAssignment.employee))
        .filter_by(task_id=task_id)
        .all()
    )
    return [a.to_dict_with_employee() for a in assignments]


def list_all_assignments():
    assignments = (
        TaskAssignment.query.options(
            joinedload(TaskAssignment.task), joinedload(TaskAssignment.employee)
        )
        .order_by(TaskAssignment.assigned_at.desc())
        .all()
    )
    results = []
    for a in assignments:
        data = a.to_dict_with_employee()
        task = a.task
        data.update(
            {
                "task_title": task.title if task else None,
                "task_description": task.description if task else None,
                "task_priority": task.priority if task else None,
                "task_estimated_hours": float(task.estimated_hours)
                if task and task.estimated_hours is not None
                else None,
            }
        )
        results.append(data)
    return results
